### PR TITLE
Add top-K logprobs for gpt-oss-120b

### PR DESCRIPTION
**Pull request by djordje-tt**
### Ticket
https://github.com/tenstorrent/tt-metal/issues/34080

### Problem description
Logprobs feature currently is calculated only for sampled token. Requirement is to return it for up to top-20 most probable tokens.

### Requirements
https://github.com/tenstorrent/tt-metal/issues/34080#issuecomment-4043977701

### What's changed
  - Add top-K logprobs computation on device for gpt-oss-120b model, returning top-32 logprobs + token indices per user which is sliced later in vllm.
  - Backward compatible: old models continue using the existing single sampled-token logprob path unchanged
  - New path activated via use_topk_logprobs=True flag in model args

### Algorithm description

 - After `ttnn.sampling` selects the sampled token from the gathered top-256 values (32 per device × 8 TP devices), the new logprobs path reuses those same gathered values and indices. 
 - It runs a second `ttnn.topk` to narrow 256 → 32 (tile-aligned), then computes logprobs for all 32 tokens in a single pass using the log-softmax formula: `logprob = logit - global_max - log(global_exp_sum)`, where global stats are computed via all-gather across TP devices. The sampled token is guaranteed to be within the top-32 since it was selected from them in `ttnn.sampling` ( runs top-32 hard-coded). 
- Device returns LogProbsResult(topk_logprobs[B,32], topk_indices[B,32]) to host, where per-user trimming to the requested num_logprobs (0-20) happens either in tt-metal's transfer_logprobs_to_host for demo(standalone
  path) or in vLLM's _build_logprobs_from_topk (serving path).

New: LogProbsResult dataclass and top-K logprobs path (tt_log_probs.py)
  - LogProbsResult holds top-32 logprobs + global vocab indices on device/host
  - calculate_topk_log_probs(): computes logprobs for gathered top-32 tokens using global max/sum-exp stats
  - transfer_logprobs_to_host(): converts to per-user dicts for tt-metal standalone path
  - to_cpu(): transfers raw tensors to host for vLLM path
  - Per-user state: logprobs_enabled, num_logprobs, topk_logprobs_needed
  - use_topk_logprobs flag controls tensor allocation: new path gets topk_logprobs_output[B,32] + topk_indices_output[B,32], old path keeps mask + output_tensor. Unused tensors are None.

  Updated: TTSampling (tt_sampling.py)
  - reset_params() extended with num_logprobs and empty_slots
  - forward() conditionally calls calculate_topk_log_probs (new) or calculate_log_probs (old) based on use_topk_logprobs flag
  - Argmax path returns None for logprobs (was computing unnecessary logprobs before)
  - Changed topk_global_indices dtype from int32 to uint32

  Updated: SamplingParams / generator.py
  - Added num_logprobs: int | list[int] = 0 field
  - format_sampling_params: pads and broadcasts enable_log_probs and num_logprobs to batch size
  - reset_sampling_params: passes num_logprobs and empty_slots through

  Updated: generator.py (tt_transformers)
  - read_decode_output: handles LogProbsResult in sync/async read paths
  - process_decode_output_host: converts LogProbsResult to (topk_lp[B,32], topk_idx[B,32]) tuple for vLLM, normalizes mixed DP rank formats

  Updated: gpt-oss Model (model.py)
  - _reset_with_flag wrapper accepts **kwargs for empty_slots passthrough

  New tests:
  - 6 TG Galaxy tests in `models/common/tests/test_sampling.py `(PCC, disabled state, per-user mixed, response format, state validation, torch-vs-tt PCC)
  - `test_gpt_oss_topk_logprobs` in gpt-oss unit tests
  - `run_logprobs_tests.sh` script with --all, --sampling-module, --gpt-oss options

### Tests (To be added)
- [ ] [Galaxy demo gpt-oss](https://github.com/tenstorrent/tt-metal/actions/runs/23468580848)
- [ ] [vllm nightly with sampling - gpt-oss-120b ](https://github.com/tenstorrent/tt-metal/actions/runs/23495556998/job/68387342968)
- [ ] [local gpt-oss-120b test](passed)
- [ ] [topk_logprobs unit tests](passed)

### CODE REVIEW

File: models/tt_transformers/tt/generator.py [hunk 1]
```diff
@@ -1501,15 +1501,39 @@ def decode_forward_llama_vision(
...
+            if isinstance(lp, LogProbsResult):
+                return LogProbsResult(
+                    topk_logprobs_host=lp.topk_logprobs.cpu(),
+                    topk_indices_host=lp.topk_indices.cpu(),
+                    topk_logprobs=None,
+                    topk_indices=None,
+                )
+            return lp.cpu()
+
+        def _read_logprobs_async(lp):
```
Comment (viktorpusTT) [inline models/tt_transformers/tt/generator.py]: Please avoid code duplication by having just one method with a parameter.

**Review by viktorpusTT — APPROVED** [review]

**Review by ipotkonjak-tt — COMMENTED** [review]
looks like bad rebase

**Review by mtairum — APPROVED** [review]

**Review by roseli-TT — APPROVED** [review]

File: models/tt_transformers/tt/generator.py [hunk 2]
```diff
@@ -722,14 +723,20 @@ def prefill_forward_text(
                     logits,
                     enable_trace=False,
                 )
+                # process logprobs to call .cpu()
+                host_log_probs = None
+                if isinstance(tt_log_probs, LogProbsResult):
+                    host_log_probs = tt_log_probs.to_cpu(blocking=False)
+                elif tt_log_probs is not None:
+                    host_log_probs = tt_log_probs.cpu(blocking=False)
```
Comment (gwangTT) [inline models/tt_transformers/tt/generator.py]: Is it possible to rename `to_cpu` to `cpu` on `LogProbsResult`? It would get rid of this if-else

Comment (gwangTT) [inline models/tt_transformers/tt/generator.py]: also, what happens in the `else:` case
```suggestion
                if isinstance(tt_log_probs, LogProbsResult):
                    host_log_probs = tt_log_probs.to_cpu(blocking=False)
                elif tt_log_probs is not None:
                    host_log_probs = tt_log_probs.cpu(blocking=False)
                else:
                    # do something here as well
                    # maybe host_log_probs = None?
```

File: models/tt_transformers/tt/generator.py [hunk 3]
```diff
@@ -767,13 +774,26 @@ def prefill_forward_text(
...
+                        lp_host = tt_log_probs.topk_logprobs_host
+                        idx_host = tt_log_probs.topk_indices_host
+                        lp_torch = ttnn.to_torch(ttnn.get_device_tensors(lp_host)[0])
+                        idx_torch = ttnn.to_torch(ttnn.get_device_tensors(idx_host)[0])
+                        log_probs_host = LogProbsResult(
+                            topk_logprobs=None,
+                            topk_indices=None,
+                            topk_logprobs_host=lp_torch[0, 0, user_batch_idx, :].unsqueeze(0),
+                            topk_indices_host=idx_torch[0, 0, user_batch_idx, :].unsqueeze(0),
+                        )
```
Comment (gwangTT) [inline models/tt_transformers/tt/generator.py]: coding style: refactoring this into a function/method could increase readability -- maybe a class method on `LogProbsResult`?

File: models/tt_transformers/tt/generator.py [hunk 4]
```diff
@@ -784,7 +804,34 @@ def prefill_forward_text(
...
+                            # ttnn host tensor
+                            all_lp[i] = ttnn.to_torch(lp_vals).reshape(-1)[:DEVICE_TOP_K].float()
+                            all_idx[i] = ttnn.to_torch(idx_vals).reshape(-1)[:DEVICE_TOP_K].int()
+                return output_tokens, (all_lp, all_idx)
+            else:
+                # Old path: build flat tensor from list
+                flat_lp = torch.ones(batch_size, dtype=torch.float32)
+                for i, lp in enumerate(output_log_probs):
+                    if lp is not None and isinstance(lp, (torch.Tensor, float, int)):
+                        flat_lp[i] = float(lp)
```
Comment (gwangTT) [inline models/tt_transformers/tt/generator.py]: coding style: refactoring this into a method/function could improve readability -- maybe name the function like `reformat_logprobs`.

**Review by gwangTT — APPROVED** [review]
Approved to unblock but please consider the coding style comments.

**Review by dpopovTT — APPROVED** [review]

File: models/common/sampling/tt_log_probs.py [hunk 5]
```diff
@@ -130,15 +198,61 @@ def _perform_all_gather(self, tensor: ttnn.Tensor, dim: int, num_links: int, buf
...
+            else:
+                self.logprobs_enabled = [enable_log_probs] * self.batch_size
+
+            if num_logprobs is not None:
+                if isinstance(num_logprobs, list):
+                    self.num_logprobs = list(num_logprobs)
+                else:
+                    self.num_logprobs = [num_logprobs] * self.batch_size
+            else:
+                self.num_logprobs = [0] * self.batch_size
```
[the remark this answers is not in the document]
Comment (djordje-tt) [reply models/common/sampling/tt_log_probs.py:325]: It's already of batch_size so no needed for this.

File: models/common/sampling/tt_log_probs.py [hunk 6]
```diff
@@ -305,3 +435,215 @@ def calculate_log_probs(
...
+            results.append(
+                {
+                    "returned_token": {
+                        "token_idx": sampled_id,
+                        "logprob": sampled_logprob,
+                    },
+                    "top_logprobs": {
+                        "token_indices": user_indices[:n].tolist(),
+                        "logprobs": user_logprobs[:n].tolist(),
+                        "num_logprobs": n,
```
[the remark this answers is not in the document]
Comment (djordje-tt) [reply models/common/sampling/tt_log_probs.py:707]: Response returns 32 values and then vllm clamps it to requested count. No need to do it on metal side.

File: models/common/sampling/tt_sampling.py [hunk 7]
```diff
@@ -409,9 +423,9 @@ def forward(
                 keepdim=False,
                 use_multicore=True,
             )
-            # Return dummy log-probs tensor with same shape as regular log-probs would be
-            # to satisfy the return type and for later post-processing
-            self.tt_log_probs = self.log_probs_calculator.calculate_log_probs(x, tt_out_tok)
+            # Argmax path: logprobs not supported (force-argmax is disabled
+            # when logprobs are enabled via format_sampling_params guard).
+            self.tt_log_probs = None
             return tt_out_tok, self.tt_log_probs
```
[the remark this answers is not in the document]
Comment (djordje-tt) [reply models/common/sampling/tt_sampling.py:429]: Currently it's not required to have logprobs calculated for force_argmax path.

File: models/tt_transformers/tt/generator.py (hunk 2 above)
Comment (djordje-tt) [reply models/tt_transformers/tt/generator.py]: Well that case is already handled above setting as None but I will put to None in else block

Comment (djordje-tt) [reply models/tt_transformers/tt/generator.py]: Done!

Comment (djordje-tt) [reply models/tt_transformers/tt/generator.py]: Done!

File: models/tt_transformers/tt/generator.py (hunk 4 above)
Comment (djordje-tt) [reply models/tt_transformers/tt/generator.py]: Done!

**Review by sraizada-tt — APPROVED** [review]
